Added binary storage option

Geolykt opened the pull request:
## Pros:
This option is significantly better than the current YAML storage in my benchmarks (with 1 million players databases).
 An improvement of around 7000ms (80% of the original value) is visible for both loading and saving 1 000 000 players, while I have to agree that 1 million players are a lot and won't be seen in normal production, it's also good for smaller servers because why not have a more optimised way of storing information? 

## Contras / Limitations
The solution only supports a bit more than 80 million players, after that it breaks due to JVM specifications (however most things would break at that point either way).
Additionally it writes data way too often and this should probably be fixed as that makes it pretty sensitive against corruption (maybe some automatic backup form). The I/O performance however is always better than the YAML implementations as of now

## Other notes
On the topic of backups, the resulting file can probably not be compressed to it's fullest extent (as we store ordering semi-randomly - however there's no need for order, so compressibility should be the main concern on ordering), but that is something to be aware of later.
I did not bump the plugin version myself, but I recommend to do so

Geolykt committed:
Added binary storage option

This option is significantly better than the current YAML storage in my benchmarks (with 1 million players databases).
 An improvment of arounds 7000ms (80% of the original value) is visible for both loading and saving 1 000 000 players, while I have to agree that 1 million players are a lot and won't be seen in normal production, it's also good for smaller servers because why not have a more optimised way of storing information? It should be noted however that the solution only supports a bit more than 80 million players, after that it breaks due to JVM specifications. It should be noted that it writes data way too often and this should probably be fixed as that makes it pretty sensitive against corruption (maybe some automatic backup form). ON the topic of backups, the resulting file can probably not be compressed to it's fullest extent (as we store ordering semi-randomly - however there's no need for order, so compressionabillity should be the main concern on ordering), but that is something to be aware of later.

Signed-off-by: Geolykt <admin@geolykt.de>

WeiiswurstDev commented:
Thanks a lot for your contribution!
I will look into your changes tomorrow.
I am also thinking about adding a way to properly credit contributors for their help, as of right now, if I merge the PR you will simply be added to the message in /eco (sth like "thanks to Geolykt for the binary storage option") and you will appear in the contributors on github.

Geolykt commented:
I've just looked at the code again and found a critical issue that would need to be fixed before it can be pulled.

Geolykt committed:
Tiny bugfix

Accidentally reveresed the orders of the UUID longs, that has been fixed

Signed-off-by: Geolykt <admin@geolykt.de>

Geolykt commented:
Aight, fixed that bug, however I'm still unable to test the plugin as I get 
```
[10:08:00 INFO]: [EasyConomy] [TRANSFER] 063ff478-dbc5-4a4a-a185-fb6ce69f1542 $2500.0
[10:08:00 INFO]: Recalculating top balances (If you have a lot of accounts, this should happen very rarely)
[10:08:00 INFO]: [EasyConomy] Write to 063ff478-dbc5-4a4a-a185-fb6ce69f1542: 2500.0 and now saving.
[10:08:00 INFO]: [EasyConomy] Storage file balances.dat saved within 1ms.
[10:08:00 WARN]: [EasyConomy] Task #582 for EasyConomy v1.3a generated an exception
org.bukkit.command.CommandException: Unhandled exception executing command 'eco' in plugin EasyConomy v1.3a
	at org.bukkit.command.PluginCommand.execute(PluginCommand.java:47) ~[patched_1.16.4.jar:git-Paper-267]
	at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:159) ~[patched_1.16.4.jar:git-Paper-267]
	at org.bukkit.craftbukkit.v1_16_R3.CraftServer.dispatchCommand(CraftServer.java:802) ~[patched_1.16.4.jar:git-Paper-267]
	at org.bukkit.Bukkit.dispatchCommand(Bukkit.java:749) ~[patched_1.16.4.jar:git-Paper-267]
	at dev.wwst.easyconomy.events.JoinEvent.lambda$onJoin$0(JoinEvent.java:25) ~[?:?]
	at org.bukkit.craftbukkit.v1_16_R3.scheduler.CraftTask.run(CraftTask.java:99) ~[patched_1.16.4.jar:git-Paper-267]
	at org.bukkit.craftbukkit.v1_16_R3.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:468) ~[patched_1.16.4.jar:git-Paper-267]
	at net.minecraft.server.v1_16_R3.MinecraftServer.b(MinecraftServer.java:1297) ~[patched_1.16.4.jar:git-Paper-267]
	at net.minecraft.server.v1_16_R3.DedicatedServer.b(DedicatedServer.java:377) ~[patched_1.16.4.jar:git-Paper-267]
	at net.minecraft.server.v1_16_R3.MinecraftServer.a(MinecraftServer.java:1212) ~[patched_1.16.4.jar:git-Paper-267]
	at net.minecraft.server.v1_16_R3.MinecraftServer.w(MinecraftServer.java:1000) ~[patched_1.16.4.jar:git-Paper-267]
	at net.minecraft.server.v1_16_R3.MinecraftServer.lambda$a$0(MinecraftServer.java:177) ~[patched_1.16.4.jar:git-Paper-267]
	at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.lang.IndexOutOfBoundsException: No group 2
	at java.util.regex.Matcher.start(Matcher.java:483) ~[?:?]
	at java.util.regex.Matcher.appendExpandedReplacement(Matcher.java:1091) ~[?:?]
	at java.util.regex.Matcher.appendReplacement(Matcher.java:998) ~[?:?]
	at java.util.regex.Matcher.replaceFirst(Matcher.java:1408) ~[?:?]
	at java.lang.String.replaceFirst(String.java:2081) ~[?:?]
	at dev.wwst.easyconomy.utils.MessageTranslator.getMessageAndReplace(MessageTranslator.java:95) ~[?:?]
	at dev.wwst.easyconomy.commands.EcoCommand.performOperation(EcoCommand.java:92) ~[?:?]
	at dev.wwst.easyconomy.commands.EcoCommand.onCommand(EcoCommand.java:53) ~[?:?]
	at org.bukkit.command.PluginCommand.execute(PluginCommand.java:45) ~[patched_1.16.4.jar:git-Paper-267]
	... 12 more
```
Every time I join. I've discarded it as my responsibility though as the stack-trace doesn't go all too far inside classes and methods that I modified in this PR.
Additionally all other commands are borked for the same (or similar) reason.

WeiiswurstDev commented:
I encountered that error as well, in my case it happened because there were players in the storage file that never joined the server before.
To fix it, I added a small error handler in the MessageTranslator, just update your fork to https://github.com/WeiiswurstDev/EasyConomy/commit/9041a84ccfcd590a0620a7b972009d2bb6d5755c and it should work.

I also noticed that you technically added support to store bank accounts, but never implemented that feature in the EasyConomyProvider.

Geolykt commented:
Cool, the Support for bank accounts was intended for use in the geolykt-prod branch (and was committed to there) so you probably checked out the wrong branch. I probably won't PR the branch myself though as It contains some significant style changes (i. e. removed static abuse in favour of dependency injection and some other stuff), however feel free to PR it yourself.

Geolykt commented:
Aright, force pushed the changes, I'll test out whether it actually works now

Geolykt commented:
Found a fix for it, it's a bit sloppy but works for me (will commit it into here).

Geolykt committed:
Fixed Exception when invoking commands

This fix is probably a bit sloppy but hey, it works. Also saved a few lines of source code, because why not

WeiiswurstDev commented:
> Cool, the Support for bank accounts was intended for use in the geolykt-prod branch (and was committed to there) so you probably checked out the wrong branch

Yes, IDEA didn't accept the url with path-1 at the end so I just removed it, I guess I cloned the wrong thing lol
> Found a fix for it, it's a bit sloppy but works for me (will commit it into here).

The fix is great, I can't believe I didn't know about the format method (or rather that it could be used in that way).

Now some smaller things that I found:
(Warning: Nit-picking follows)

https://github.com/Geolykt/EasyConomy/blob/path-1/src/main/java/dev/wwst/easyconomy/storage/BinaryDataStorage.java#L153
Why would you get the logger from `EasyConomy.getInstance().getLogger()` if you could also use `plugin.getLogger()`? (This happens a few times)

I went through the code in a little more detail and I found that you would shutdown the entire server if the storage file might be corrupted: https://github.com/Geolykt/EasyConomy/blob/path-1/src/main/java/dev/wwst/easyconomy/storage/BinaryDataStorage.java#L160

I would rather just disable the plugin, the damage will be prevented either way and that way you don't interfere with the other plugins (that's also how I did it in the onEnable method if something goes wrong).

Anyways thanks a lot for all the work you are putting in this pr, I'm really thankful for it 👍

Geolykt committed:
Implement proposed nitpicks

This basically gets rid of useless (or better said uncached) calls and doesn't do much outside of that.

Signed-off-by: Geolykt <admin@geolykt.de>

Geolykt commented:
Changed. I also throw an IOException since I think it's appropriate and it's not really specified when a plugin terminates when the Bukkit.disablePlugin(plugin) is called, so I just made sure in this sense.

WeiiswurstDev commented:
Great, will merge soon™ (will be in the next update, whenever that will be)
You will certainly be mentioned in the credits (/eco) and on the Github, thanks again for your contribution!
